### DISCUSSION TITLE
Blazor SS and ApplicationContext DI weirdness and feedback

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#2760] [Questions] [answered]
@rockfordlhotka  @TheCakeMonster 
(Not sure if this is a bug or me doing something wrong - so posting several questions below instead of new issue).  At the risk of adding 11th hour stuff...

It might help me test if you could spare a few minutes to put out one more pre-release of CSLA 6.  That would help me rule out any errors in my creation of local nuget packages.  I will gladly test against that and report back what I find.  

It seems I might be one of the few testing a Blazor SS (only) project, but it's turning up interesting things.  I'm testing your latest changes.  The project has no reference to `Csla.AspNetCore`.  It only references `Csla` and `Csla.Blazor.`  

Here is the relevant line in program.cs:
`builder.Services.AddCsla(options => options
    .AddServerSideBlazor());
`

I did the following:
1. pulled latest csla Repo
2. (since you haven't done another preview) I manually made local only versions of nuget packages for csla.nupkg and csla.blazor.nupkg
3. Changed my business object factory constructors to inject `Csla.IDataPortalFactory` instead of the old way of injecting typed data portals.

When I hit the following line "Portal = ..." :
```cs
public class BetaInfoListFactory : DemoReadOnlyListBaseFactory<BetaInfoList, BetaInfo>
    {
        public BetaInfoListFactory(Csla.IDataPortalFactory dpFactory, ApplicationContext applicationContext)
        {
            Portal = dpFactory.GetPortal<BetaInfoList>();
            ...
```

I get this error:
```A suitable constructor for type 'Csla.IDataPortal`1[BusinessLayer.BetaInfoList]' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.```

**Issue/Question 1:**  
I am unable to get past that error no matter what I try in the `builder.Services` configuration in Program.cs.

**Issue/Question 2:**  
If I debug and breakpoint on that same line...I notice the `ApplicationContext.ContextManager` is of type `{Csla.Blazor.WebAssembly.ApplicationContextManager}.`  That is freaky because I ONLY have Csla and Csla.Blazor packages referenced.  I have no reference to Csla.Blazor.WebAssembly.  So where did it pull that from?  I do NOT have any call to `WithBlazorWebAssembly` in my program.cs.  I would have expected it to get the default one from the Csla project. 
Note:  I tried adding a reference to Csla.AspNetCore and it still picks up the WebAssembly one!  Weird.

**Issue/Question 3:**  
I see you removed the Csla.Blazor only version of ApplicationContextManager from `Csla.Blazor` and instead you  put it in `Csla.AspNetCore.Blazor`. IMHO it feels off that Csla requires me to know that I should also reference Csla.AspNetCore in a BlazorSS only project.  I realize you would end up having two versions of the Blazor.Server ApplicationContextManager (one for BSS only and one for AspNetCore).  
Perhaps a little code duplication is appropriate - maybe "Add existing file via link" could fix that.   
OR
If you are expecting us BSS only peeps to always reference Csla.AspNetCore...then wouldn't it need to be a dependency in the Nuget package so we don't have to remember.

**Minor Issue 4:**  
The comment for the method `Csla.Blazor.WebAssembly.BlazorWasmConfigurationExtensions.WithBlazorWebAssembly` errantly says the following:
> Registers services necessary for Windows Forms

**Suggestion 5:**
Would you consider having Csla Samples include a simpler BlazorSS only project which does not have any extra overhead.  The current one is a bit confusing because it is mixing client (WASM) and server projects, and because the server project references AspNetCore.

**Comment by rockfordlhotka** [comment, marked as answer]
For server-side Blazor you will be required to reference the `Csla.AspNetCore` and `Csla.Blazor` packages. I had to put the authorization types into `Csla.AspNetCore` because they rely on types that aren't referenced by `Csla.Blazor` (and can't be, because it is also used in Blazor wasm).

As I'm updating the Blazor sample app I'm fixing some bugs in the configuration API. Possibly regressions, but still, bugs.

Good idea about a server-side Blazor example to go with the existing Blazor wasm one.

I will push another prerelease once I have at least a couple of the sample apps running. No sense pushing it now, as I know there are bugs that I'm actively fixing.

**Reply by swegele** [reply]
Would it be possible to have the `Csla.Blazor` package simply reference the `Csla.AspNetCore` package so that users are clear it's required rather than wondering and finding out only at run-time?

I will wait for the next pre-release to be sure that I haven't stumbled on a bug you've already fixed.
Thanks!

**Reply by rockfordlhotka** [reply]
The `Csla.Blazor` project needs to support client-side Blazor, which doesn't work with server-side aspnetcore dependencies, so it can't auto-reference the server-side components.

**Reply by rockfordlhotka** [reply]
I'm pretty sure this is the cause of one of your issues 😳

https://github.com/MarimerLLC/csla/pull/2761/commits/bd5652ec4a07ba622bb0f31995dd2a54070f95b3

**Reply by swegele** [reply]
Looking at your Blazor sample, I saw the `BlazorExample.Client` only references `Csla.Blazor.WebAssembly`.  So I assumed (keyword there) that `Csla.Blazor` was not related at all to WASM.   Clearly there are more scenarios than what I am aware of.

**Reply by swegele** [reply]
DUHHH I see it now...I should have expanded the `Csla.Blazor.WebAssembly` package.

**Reply by swegele** [reply]
I can't help but keep thinking there is a design issue here that is causing friction (maybe just a fact of life).

If user is building Blazor WASM then all is well...the packages match and support what is expected (Csla.Blazor.WebAssembly package is dependent on Csla.Blazor)  But if user is doing Blazor.Server, they must have insider knowledge that Csla.AspNetCore is needed.

Almost seems like we are missing a Csla.Blazor.Server and a Common.

So?  While I'm offering arm-chair-quarterback free advice...

Csla.Blazor.Server  
  |__>  Csla.Blazor.Common
  |__>  Csla.AspNetCore

Csla.Blazor.WebAssembly  
  |__>  Csla.Blazor.Common

No need to respond...this is either totally dumb or the best idea ever and I trust you to recognize which :-)

**Reply by rockfordlhotka** [reply]
The way I think about it is:

1. If you are running in aspnetcore (for any reason), then you need the `Csla.AspNetCore` package
2. If you are building a Blazor app (of any sort), then you need the `Csla.Blazor` package
3. If you are building a Blazor wasm app, then you need the `Csla.Blazor.WebAssembly` package

This generally follows the model Microsoft has provided for these technologies.

**Comment by swegele** [comment]
Totally agree functionally (now that I’ve learned about how it all works).
Indeed a true Blazor SS app STILL has .cshtml pages _Host and _Layout etc
for initial hookup at least.

Just was trying to think of a way to get the naming to be easiest to
understand and for the prerequisites to automatically match while avoiding
the reference issues that currently stop Cala.AspNetCore from being a
prerequisite of Cala.Server.  Maybe it was just me but the way it is now
was not intuitive for me and I had the impression that adding
Csla.AspNetCore was adding unnecessary stuff.

But perhaps clear documentation and the BSS only sample would sufficiently
help those that come later over the years.
